### ISSUE TITLE
StateDirectory= and CacheDirectory= are not chowned to User=

### ISSUE THREAD

**Issue by fsateler** [systemd/systemd#11842]
**systemd version the issue has been seen with**
> 240, 241

**Used distribution**
 > Debian

**Expected behaviour you didn't see**
 > CacheDirectory and StateDirectory are chowned to the user specified in User before launching the service

**Unexpected behaviour you saw**
 > CacheDirectory and StateDirectory are not chowned if they have the wrong owner. FooDirectoryMode= isn't changed as well

**Steps to reproduce the problem**

```
% systemctl cat test-state
# /etc/systemd/system/test-state.service
[Unit]

[Service]
Type=oneshot
StateDirectory=test
CacheDirectory=test
RuntimeDirectory=test
User=1000
ExecStart=/bin/ls -ld $RUNTIME_DIRECTORY $STATE_DIRECTORY $CACHE_DIRECTORY

% sudo systemctl start test-state ; journalctl -u test-state -n
-- Logs begin at Wed 2018-09-26 12:31:05 -03, end at Wed 2019-02-27 09:20:16 -03. --
Feb 27 09:20:16 felipedell systemd[1]: Starting test-state.service...
Feb 27 09:20:16 felipedell ls[7203]: drwxr-xr-x 2 felipe felipe   40 Feb 27 09:20 /run/test
Feb 27 09:20:16 felipedell ls[7203]: drwxr-xr-x 2 felipe felipe 4096 Feb 27 09:20 /var/cache/test
Feb 27 09:20:16 felipedell ls[7203]: drwxr-xr-x 2 felipe felipe 4096 Feb 27 09:20 /var/lib/test
Feb 27 09:20:16 felipedell systemd[1]: test-state.service: Succeeded.
Feb 27 09:20:16 felipedell systemd[1]: Started test-state.service.
% sudo chown root:root /var/cache/test /var/lib/test 
% sudo systemctl start test-state ; journalctl -u test-state -n
Feb 27 09:20:43 felipedell systemd[1]: Starting test-state.service...
Feb 27 09:20:43 felipedell ls[7679]: drwxr-xr-x 2 felipe felipe   40 Feb 27 09:20 /run/test
Feb 27 09:20:43 felipedell ls[7679]: drwxr-xr-x 2 root   root   4096 Feb 27 09:20 /var/cache/test
Feb 27 09:20:43 felipedell ls[7679]: drwxr-xr-x 2 root   root   4096 Feb 27 09:20 /var/lib/test
Feb 27 09:20:43 felipedell systemd[1]: test-state.service: Succeeded.
Feb 27 09:20:43 felipedell systemd[1]: Started test-state.service.
```

This appears to have been broken before 239 ( see #11128 ), but not properly fixed.

**Comment by taro-yamada** [comment]
I think it is an intended behavior that "not changing the owner, group or mode if the {Cache,Logs,State}directory already exist (and dynamicuser is false)". 
30c81ce2cef97b05e143c7adf4cd1b1c5fb59932
(though it may need more discussion about its behavior...)


It is my fault that it only warn about the difference in mode.
I want to send PR if "just warning" is right way to fix this issue.

**Comment by mbiebl** [comment]
@taro-yamada if it is intentional, that those *Directory= settings behave differently for the dynamic user case, then this should be reflected in the documentation. It currently says
```
           Except in case of ConfigurationDirectory=, the innermost specified
           directories will be owned by the user and group specified in User=
           and Group=. If the specified directories already exist and their
           owning user or group do not match the configured ones, all files
           and directories below the specified directories as well as the
           directories themselves will have their file ownership recursively
           changed to match what is configured. As an optimization, if the
           specified directories are already owned by the right user and
           group, files and directories below of them are left as-is, even if
           they do not match what is requested. The innermost specified
           directories will have their access mode adjusted to the what is
           specified in RuntimeDirectoryMode=, StateDirectoryMode=,
           CacheDirectoryMode=, LogsDirectoryMode= and
           ConfigurationDirectoryMode=.
```

**Comment by poettering** [comment]
yeah, i guess whether dynamicuser is used or not shouldn't matter, what matters is that the service is run with a user different from the service managers, and that could be either a static or a dynamic one, the effect via chown()ing should be the same

**Comment by poettering** [comment]
Fix pending in #12005
